scipy/optimize/_trustregion_iterative.py

```diff
@@ -0,0 +1,470 @@
+
+    Returns
+    -------
+    delta : float
+        Amout that should be added to the element (k, k) of the
+        leading k by k submatrix of ``a`` to make it singular.
+    v : ndarray
+        A vector such that ``v.T B v = 0``. Where B is the matrix A after
+        ``delta`` is added to its element (k, k).
+    """
+
+    # Compute delta
+    delta = np.sum(c[:k-1, k-1]**2) - a[k-1, k-1]
+
+    # Get dimension
+    n = len(a)
+
+    # Inicialize v
+    v = np.zeros(n)
+    v[k-1] = 1
+
+    # Compute the remaining values of v by solving a triangular system
+    if k != 1:
+        v[:k-1] = solve_triangular(c[:k-1, :k-1], -c[:k-1, k-1])
+
+    return delta, v
+
+
+class IterativeSubproblem(BaseQuadraticSubproblem):
+    """Quadratic subproblem solved by the iterative method.
+
+    Notes
+    -----
+    This subproblem solver was based on _[1], _[2] and _[3],
+    which implement similar algorithms. The algorithm is basically
+    that of _[1] but ideas from _[2] and _[3] were also used.
+
+    References
+    ----------
+    .. [1] Conn, A. R., Gould, N. I., & Toint, P. L. (2000).
+           "Trust region methods". Siam. pp. 169-200.
+    .. [2] Nocedal, J., and  Wright, S. (2006).
+           "Numerical optimization". Springer Science &
+           Business Media. pp. 83-91.
+    .. [3] Moré, J. J., & Sorensen, D. C. (1983).
+           Computing a trust region step. SIAM Journal on Scientific
+           and Statistical Computing, 4(3), 553-572.
+    """
+
+    UPDATE_COEF = 0.01
+    EPS = np.finfo(float).eps
+
+    def __init__(self, x, fun, jac, hess=None, hessp=None,
+                 optimization_factor=0.8, trust_region_factor=1.1):
+
+        self.lb = None
+        self.previous_trust_radius = -1
+        self.niter = 0
+        self.easycase_stop_criteria = min(trust_region_factor-1,
+                                          1-np.sqrt(optimization_factor))
+        self.hardcase_stop_criteria = 1-optimization_factor
+        self._hess_gershgorin_ub = None
+        self._hess_gershgorin_lb = None
+        self._hess_inf = None
+        self._hess_frob = None
+        self._dimension = None
+
+        # Superclass __init__
+        BaseQuadraticSubproblem.__init__(self, x, fun, jac,
+                                         hess)
+
+        # Get Lapack function for cholesky decomposition
+        self.cholesky, = get_lapack_funcs(('potrf',), (self.hess,))
+
+        # A constant such that for vectors smaler than that
+        # backward substituition is not reliable. It was stabilished
+        # based on Golub, G. H., Van Loan, C. F. (2013).
+        # "Matrix computations". Forth Edition. JHU press., p.165.
+        self.CLOSE_TO_ZERO = self.dimension*self.EPS*self.hess_inf
+
+    @property
+    def dimension(self):
+        """Hessian inf norm"""
+        if self._dimension is None:
+            self._dimension = len(self.hess)
+        return self._dimension
+
+    @property
```

Comment:
Actually now I realize that `hess_gershgorin_ub` (and so on) can be reused when `solve` is called multiple times. So just recomputing looks suboptimal. 

Perhaps properties are fine or maybe it's better to move everything to the constructor. Please consider both options and use your best judgment. I just want that everything is as straightforward as possible.